### DISCUSSION TITLE
Duda ejercicio backtracking

### DISCUSSION THREAD

**Discussion by IgnacioHqz** [IIC2133-PUC/2024-2#182] [Preguntas Materia] [answered]
Hola, estaba viendo el ejercicio que se hizo en clases sobre backtracking, el ejercicio de la barcaza, pero no logro entender por qué se le va restando a L el largo de cada auto, y a W el peso de cada auto si es que se sube a la barcaza. Eso no afectaría a la restricción? Entiendo que se está restando para actualizar el largo disponible que queda por asignar, pero no sería mejor utilizar una variable que vaya sumando los largos y pesos y que vaya viendo si es menor a L y a W o debe ser consistente con lo que se colocó en las restricciones? No sé si se entiende mi duda.
Muchas gracias!

Este es el ejercicio:

![image](https://github.com/user-attachments/assets/66eb644a-4140-4ebe-b53a-a541de873f97)

**Comment by Aleksandsk** [comment, marked as answer]
![image](https://github.com/user-attachments/assets/bbc86e15-c981-4921-aecb-cfe9d59577ed)

Efectivamente el código no es totalmente consistente con la restricción que se plantea unas diapositivas antes en la parte a), entonces como en cada llamada a Backtrack se resta l(i) a L y  w(i) a W cuando j = 1, al momento de revisar si X[i] = j supera las restricciones, habría que verificar que j*l(i) <= L y j*w(i) <= W, ya que L y W han sido reducidas en las llamadas anteriores. Con eso estas verificando si queda espacio y capacidad para el auto i dado que ya se asignaron cierto autos anteriores a i.

Las restricciones en la parte a) son correctas, lo que ocurre en el código es que falta especificar como se verifica que se cumplan. En este caso hacer lo que dije en el párrafo anterior es equivalente a no reducir W y L en las llamadas de Backtrack y comprobar las restricciones (desigualdades) tal y como están escritas en a). Con eso el código seria consistente con las restricciones en a) pero seria menos eficiente.

La idea que propones tu también es valida y en la prueba hay que especificar como tu implementación cumple las restricciones, ya sea con un metodo check() o escribiendo directamente las restricciones en el código de forma que tengan sentido.

Espero que haya quedado mas claro!

**Reply by IgnacioHqz** [reply]
Perfecto! Pero entonces las restricciones que se escriben en a) deben ser las mismas que se utilicen en la parte b) o puedo modificar las restricciones que colocaré en b) ?

Lo que no me queda claro es que siguiendo la idea que me comentas de chequear si l(i)* j <= L, en ese caso L significaría el espacio que queda en la barcaza no? pero si es así entonces la letra L que se colocó en las restricciones ya no significaría el largo de la barcaza. Cómo se tendría que hacer ahí?

**Reply by Aleksandsk** [reply]
Respecto a tu primera pregunta, la idea es respetar la capacidad y el espacio máximo de la barcaza, entonces si en la parte (a) representas estas restricciones con una desigualdad (y están correctas) y en la parte de la implementación no utilizas exactamente esas desigualdades esta bien mientras cumplas la idea original de respetar el espacio y las capacidades.

En cuanto a tu segunda pregunta, efectivamente L significaría el espacio que queda en la barcaza (análogamente peso disponible para W), entonces al principio en la primera llamada a Bactrack, tienes todos los autos posibles para asignar y espacio original L, luego en cada llamada recursiva a Backtrack, lo que estas haciendo ahora es resolver un problema mas chico, ya que te quedan menos autos por asignar y menos capacidad disponible que llamaremos L'. Cuando llegas a que l(i)* j > L', quiere decir que ese auto i no puede ingresar porque supera el espacio que queda disponible, y debido a que el problema mas chico depende de las asignaciones anteriores, esto es lo mismo que decir que al sumar el largo de ese auto i con la suma de los largos de todos los autos anteriores que les asignaste j = 1, la sumatoria superara al L original. Entonces al final ir disminuyendo los L' termina de igual forma comprobando que se cumpla el L original.

**Reply by IgnacioHqz** [reply]
perfecto! Muchas gracias por la explicación! Me había olvidado de agradecerte
Ahora quedó más claro, saludos!!
